content/advent-2016/some-tools-for-go-that-you-might-not-know-yet.md

```diff
@@ -0,0 +1,221 @@
++++
+author = ["Christoph Berger"]
+date = "2016-12-10T00:00:00+01:00"
+title = "Some Tools For Go That You Might Not Know Yet"
+series = ["Advent 2016"]
+draft = true
++++
+
+Year's end is coming closer. Time to clean up repositories and 
+polishing up the toolset. All the well-known tools have been installed
+already--is there anything else to add to the toolbox?
+
+Here are a few useful tools that you might not have in your toolbox yet:
+`interfacer`, `zb`, `realize`, and `binstale`. They have nothing in
+common except that each of them solves a particular problem well. 
+
+
+
+# *interfacer:* Should I rather use an interface here?
+
+`interfacer` has a very specific
+purpose: It looks at the parameters of your functions and points out
+those that you could replace by an interface type. 
+
+Why this? 
+
+Maybe you have heard of the following advice: A function should expect
+an interface and return a specific type. I fail to remember where I came
+across that rule, and the precise wording might also be different, but
+the point is that if a function parameter is an interface then the
+function is much more versatile and can, for example, receive a mock
+type when used in a unit test. 
+
+So whenever you feel you have missed an opportunity to have one of your
+functions receive an interface rather then a struct, run `interfacer` 
+and it will tell you if you did.
+
+## An example
+
+Imagine that one day, you start writing a larger library that contains a
+`Doer` interface and a `DoIt` struct that implements `Doer`.  
+
+
+```go 
+type Doer interface { 
+	Do() 
+}
+
+
+type DoIt struct {
+	//...	
+}
+
+func (DoIt) Do() {
+	//...
+}
+```
+
+A couple hours and few thousand lines later (yes, you feel productive 
+this day) you define a function `DoSomething` that expects a `DoIt` 
+struct and calls its `Do` method.
+
+```go 
+func DoSomething(d DoIt) {
+	d.Do()
+}
+```
+
+You vaguely remember that `Do` belongs to some interface but you can't 
+remember which one. You're too lazy to search for it (and after all, 
+it is already 11pm), so you run 
+
+    $ interfacer dosomething.go
+
+and get this advice:
+
+    dosomething.go:12:18: d can be Doer
+
+You quickly fix the `DoSometing` function and go to bed, knowing that
+you now can easily pass some `MockDoIt` struct to `DoSomething` when
+you'll write the tests tomorrow.
+
+[Interfacer on GitHub](https://github.com/mvdan/interfacer/)
+
+
+# *zb:* Take some shortcuts to the go toolchain
+
+After you finished some work on your latest project, you run
+gometalinter. It takes some time to finish, and you discover that some
```

Comment:
I added code formatting to gometalinter and to all other commands I found, for consistency. I was afraid that this could create some "visual clutter" from too many code formatting in the text, but I think the result is still quite ok.